src/Cryptol/Utils/RecordMap.hs

```diff
@@ -0,0 +1,181 @@
+instance (Show a, Ord a, Show b) => Show (RecordMap a b) where
+  show = show . displayFields
+
+instance (NFData a, NFData b) => NFData (RecordMap a b) where
+  rnf = rnf . canonicalFields 
+
+-- | Generate a record from a list of field/value pairs.
+--   Precondition: each field identifier appears at most
+--   once in the given list.
+recordFromFields :: (Show a, Ord a) => [(a,b)] -> RecordMap a b
+recordFromFields xs =
+  case recordFromFieldsErr xs of
+    Left (x,_) -> 
+          panic "recordFromFields"
+                ["Repeated field value: " ++ show x]
+    Right r -> r
+
+-- | Generate a record from a list of field/value pairs.
+--   If any field name is repeated, the first repeated name/value
+--   pair is returned.  Otherwise the constructed record is returned.
+recordFromFieldsErr :: (Show a, Ord a) => [(a,b)] -> Either (a,b) (RecordMap a b)
+recordFromFieldsErr xs0 = loop mempty xs0
+  where
+  loop m [] = Right (RecordMap m (map fst xs0))
+  loop m ((x,v):xs)
+    | Just _ <- Map.lookup x m = Left (x,v)
+    | otherwise = loop (Map.insert x v m) xs
+
+-- | Generate a record from a list of field/value pairs,
+--   and also provide the "display" order for the fields directly.
+--   Precondition: each field identifier appears at most once in each
+--   list, and a field name appears in the display order iff it appears
+--   in the field list.
+recordFromFieldsWithDisplay :: (Show a, Ord a) => [a] -> [(a,b)] -> RecordMap a b
+recordFromFieldsWithDisplay dOrd fs = r { displayOrder = dOrd }
+  where
+  r = recordFromFields fs
+
+-- | Lookup the value of a field
+lookupField :: Ord a => a -> RecordMap a b -> Maybe b
+lookupField x m = Map.lookup x (recordMap m)
+
+
+-- | Update the value of a field by applying the given function.
+--   If the field is not present in the record, return Nothing.
+adjustField :: forall a b. Ord a => a -> (b -> b) -> RecordMap a b -> Maybe (RecordMap a b)
+adjustField x f r = mkRec <$> Map.alterF f' x (recordMap r)
+  where
+  mkRec m = r{ recordMap = m }
+
+  f' :: Maybe b -> Maybe (Maybe b)
+  f' Nothing = Nothing
+  f' (Just v) = Just (Just (f v))
+
+-- | Return a list of field/value pairs in canonical order
+canonicalFields :: RecordMap a b -> [(a,b)]
+canonicalFields = Map.toList . recordMap
+
+-- | Return a list of field/value pairs in display order
+displayFields :: (Show a, Ord a) => RecordMap a b -> [(a,b)]
+displayFields r = map find (displayOrder r)
+  where
+  find x =
+    case Map.lookup x (recordMap r) of
+      Just v -> (x, v)
+      Nothing ->
+         panic "displayFields"
+               ["Could not find field in recordMap " ++ show x
+               , "Display order: " ++ show (displayOrder r)
+               , "Canonical order:" ++ show (map fst (canonicalFields r))
+               ]
+
+traverseRecordMap :: Applicative t =>
+  (a -> b -> t c) -> RecordMap a b -> t (RecordMap a c)
+traverseRecordMap f r =
+  (\m -> RecordMap m (displayOrder r)) <$> Map.traverseWithKey f (recordMap r)
```

Comment:
It might be nice to add a pure variant of this function, akin to `mapWithKey` from `Data.Map`. This would simplify those places where you use `traverseRecordMap` with the `Identity` functor.